### PR TITLE
Spack recipes for AOCL 3.0

### PR DESCRIPTION
**Pull request by amd-toolchain-support**
(1) blis - changed configure function with configure_args name which
    will return args to configure function

(2) amdblis - reused configure_args from blis package using super
    function and updated the amdblis specific flags for 3.0 version

(3) amdlibflame - Enabling '--enable-f2c-dotc' flags from version 3.0
    onwards

(4) amdlibm - Added reference to 3.0 release branch

(5) amdscalapack - Added reference to 3.0 release

(6) aocl-sparse - Added suport for ilp64 variant from version 3.0
   Incorporated review comments from previous below pull request
   https://github.com/spack/spack/pull/22276

### CODE REVIEW

File: var/spack/repos/builtin/packages/amdscalapack/package.py [hunk 1]
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
         args = super(Amdscalapack, self).cmake_args()
         spec = self.spec
 
-        args.extend([
-            "-DUSE_DOTC_WRAPPER:BOOL=%s" % (
-                'ON' if '%aocc ^amdblis' in spec else 'OFF'
-            )
-        ])
+        if "%gcc@10:" in spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdscalapack/package.py]:
```suggestion
        if spec.satisfies('%gcc@10:'):
```

File: var/spack/repos/builtin/packages/amdscalapack/package.py [hunk 2]
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
...
+            args.extend(['-DCMAKE_Fortran_FLAGS={0}'.format(
+                        "-fallow-argument-mismatch")])
+
+        if "@2.2" in spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdscalapack/package.py]:
```suggestion
        if spec.satisfies('@2.2'):
```

File: var/spack/repos/builtin/packages/amdlibflame/package.py [hunk 3]
```diff
@@ -56,6 +55,12 @@ def configure_args(self):
         """configure_args function"""
         args = super(Amdlibflame, self).configure_args()
         args.append("--enable-external-lapack-interfaces")
+
+        """To enabled Fortran to C calling convention for
+        complex types when compiling with aocc flang"""
+        if "@3.0: %aocc" in self.spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdlibflame/package.py:61]:
```suggestion
        if self.spec.satisfies('@3.0: %aocc'):
```

File: var/spack/repos/builtin/packages/amdblis/package.py [hunk 4]
```diff
@@ -17,9 +17,21 @@ class Amdblis(BlisBase):
...
     maintainers = ['amd-toolchain-support']
 
+    version('3.0', sha256='ac848c040cd6c3550fe49148dbdf109216cad72d3235763ee7ee8134e1528517')
     version('2.2', sha256='e1feb60ac919cf6d233c43c424f6a8a11eab2c62c2c6e3f2652c15ee9063c0c9')
+
+    def configure_args(self):
+        spec = self.spec
+        args = super(Amdblis, self).configure_args()
+
+        if "@3.0%aocc@3.0.0" in spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdblis/package.py]:
```suggestion
        if spec.satisfies('@3.0%aocc@3.0.0'):
```

File: var/spack/repos/builtin/packages/amdscalapack/package.py [hunk 5]
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
...
+            args.extend(['-DUSE_DOTC_WRAPPER:BOOL=%s' % (
+                        'ON' if '%aocc ^amdblis' in spec else 'OFF')])
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdscalapack/package.py]:
```suggestion
                        'ON' if spec.satisfies('%aocc ^amdblis') else 'OFF')])
```

File: var/spack/repos/builtin/packages/aocl-sparse/package.py [hunk 6]
```diff
@@ -64,18 +67,19 @@ def cmake(self, spec, prefix):
...
-        args.extend([
             "-DBUILD_CLIENTS_BENCHMARKS:BOOL=%s" % (
-                'ON' if self.run_tests else 'OFF'
-            )
+                'ON' if self.run_tests else 'OFF')
         ])
 
-        with working_dir(self.build_directory, create=True):
-            inspect.getmodule(self).cmake(*args)
+        if "@3.0:" in spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/aocl-sparse/package.py]:
```suggestion
        if spec.satisfies('@3.0:'):
```

File: var/spack/repos/builtin/packages/aocl-sparse/package.py [hunk 7]
```diff
@@ -64,18 +67,19 @@ def cmake(self, spec, prefix):
...
+            args.extend([
+                "-DBUILD_ILP64:BOOL=%s" % (
+                    'ON' if '+ilp64' in spec else 'OFF')
+            ])
+
+        return args
 
     # Check that self.prefix is there after installation
     @run_after('build')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/aocl-sparse/package.py:84]: FWIW. Regarding the comment above, `check` is technically testing the build -- after the `build` phase and using the `self.build_directory` -- not the installation.  Testing the installation would be done after the `install` phase and would be using the `self.prefix`, which is the root directory for the package's installed files.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Made a number of minor suggested changes and comments.  

We provide more examples of checking a spec's constraints at https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints.

File: var/spack/repos/builtin/packages/amdscalapack/package.py (hunk 1 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/amdscalapack/package.py]: Incorporated review comments

File: var/spack/repos/builtin/packages/amdlibflame/package.py (hunk 3 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/amdlibflame/package.py:61]: Incorporated review comments

File: var/spack/repos/builtin/packages/amdscalapack/package.py (hunk 2 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/amdscalapack/package.py]: Incorporated review comments

File: var/spack/repos/builtin/packages/amdblis/package.py (hunk 4 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/amdblis/package.py]: Incorporated review comments

File: var/spack/repos/builtin/packages/amdscalapack/package.py (hunk 5 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/amdscalapack/package.py]: Incorporated review comments

File: var/spack/repos/builtin/packages/aocl-sparse/package.py (hunk 7 above)
Comment (vkallesh) [reply var/spack/repos/builtin/packages/aocl-sparse/package.py:84]: Removed following comment:
# Check that self.prefix is there after installation

Changed the following text from
Simple test to test the installation by running
to
Simple test to test the built library by running

**Comment by amd-toolchain-support** [comment]
> Made a number of minor suggested changes and comments.
> 
> We provide more examples of checking a spec's constraints at https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints.

Thanks for the suggestion, all review comments are addressed.

File: var/spack/repos/builtin/packages/amdblis/package.py [hunk 8]
```diff
@@ -17,9 +17,22 @@ class Amdblis(BlisBase):
...
     maintainers = ['amd-toolchain-support']
 
+    version('3.0', sha256='ac848c040cd6c3550fe49148dbdf109216cad72d3235763ee7ee8134e1528517')
     version('2.2', sha256='e1feb60ac919cf6d233c43c424f6a8a11eab2c62c2c6e3f2652c15ee9063c0c9')
+
+    def configure_args(self):
+        spec = self.spec
+        args = super(Amdblis, self).configure_args()
+
+        if spec.satisfies('@3.0%aocc'):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdblis/package.py]:
```suggestion
        if spec.satisfies('@3.0: %aocc'):
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/amdblis/package.py]: The above assumes the code in the body of the `if` statement applies to versions >= `3.0`.

File: var/spack/repos/builtin/packages/amdscalapack/package.py [hunk 9]
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
...
-        args.extend([
-            "-DUSE_DOTC_WRAPPER:BOOL=%s" % (
-                'ON' if '%aocc ^amdblis' in spec else 'OFF'
-            )
-        ])
+        if spec.satisfies('%gcc@10:'):
+            args.extend(['-DCMAKE_Fortran_FLAGS={0}'.format(
+                        "-fallow-argument-mismatch")])
+
+        if spec.satisfies('@2.2'):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdscalapack/package.py:46]: Does this argument only apply to version `2.2`?  If not, you might want to use a range check here.  

https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints

File: var/spack/repos/builtin/packages/aocl-sparse/package.py [hunk 10]
```diff
@@ -29,8 +29,10 @@ class AoclSparse(CMakePackage):
             values=('Debug', 'Release'))
     variant('shared', default=True,
             description='Build shared library')
+    variant('ilp64', default=False,
+            description='Build with ILP64 support')
 
-    depends_on('boost')
+    depends_on('boost', when='@2.2')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/aocl-sparse/package.py:35]: Again, does this dependency only apply to version `2.2`.

File: var/spack/repos/builtin/packages/amdlibm/package.py [hunk 11]
```diff
@@ -37,8 +35,8 @@ class Amdlibm(SConsPackage):
     depends_on("scons@3.1.2:", type=("build"))
     depends_on("mpfr", type=("link"))
 
-    patch('0001-libm-ose-Scripts-cleanup-pyc-files.patch')
-    patch('0002-libm-ose-prevent-log-v3.c-from-building.patch')
+    patch("0001-libm-ose-Scripts-cleanup-pyc-files.patch", when="@2.2")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/amdlibm/package.py:38]: Is there a version before `2.2` that could be installed?  If so, then you'd want to use `@:2.2`.

File: var/spack/repos/builtin/packages/amdblis/package.py (hunk 8 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/amdblis/package.py]: These flags are specific to amdblis 3.0 version. Next version of amdblis don't need below flags. 

 if spec.satisfies('@3.0 %aocc'):
            """ To enabled Fortran to C calling convention for
            complex types when compiling with aocc flang"""
            args.append('CFLAGS={0}'.format("-DAOCL_F2C"))
            args.append('CXXFLAGS={0}'.format("-DAOCL_F2C"))

File: var/spack/repos/builtin/packages/aocl-sparse/package.py (hunk 10 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/aocl-sparse/package.py:35]: Yes, Boost is required only for aocl-sparse 2.2 version.

File: var/spack/repos/builtin/packages/amdlibm/package.py (hunk 11 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/amdlibm/package.py:38]: amdlibm 2.2 is the first version in use for spack. There is no previous version.

File: var/spack/repos/builtin/packages/amdscalapack/package.py (hunk 9 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/amdscalapack/package.py:46]: Yes, this argument only apply to 2.2 version.

**Review by tldahlgren — APPROVED** [review]
LGTM.  Approval contingent on all required checks passing.

Thanks for responding to my questions.

**Comment by tldahlgren** [comment]
Thank you for your contribution!

**Pull request merged**